Overhaul R HTML widgets and HTML support in Viewer

jmcphers opened the pull request:
This change holistically addresses a large number of issues around viewing R HTML widgets, and HTML in general, in the Viewer and Plots panes. Specifically, it addresses the following issues:

https://github.com/posit-dev/positron/issues/2023
https://github.com/posit-dev/positron/issues/2145
https://github.com/posit-dev/positron/issues/2559
https://github.com/posit-dev/positron/issues/2888
https://github.com/posit-dev/positron/issues/3018
https://github.com/posit-dev/positron/issues/3678
https://github.com/posit-dev/positron/issues/3756
https://github.com/posit-dev/positron/issues/3783
https://github.com/posit-dev/positron/issues/3914
https://github.com/posit-dev/positron/issues/3968
https://github.com/posit-dev/positron/issues/4057

Prior to this change, R HTML widgets were emitted from Ark as notebook outputs in format compatible with the [VS Code Notebook Renderer API](https://code.visualstudio.com/api/extension-guides/notebook#notebook-renderer). This approach was chosen because we need to be able to render these kinds of outputs in any case (other kernels emit them), and for minimal friction with the existing infrastructure.

However, R HTML widgets aren't really a good match for Notebook Renderer API, and the abstraction was very leaky. In particular, the API is really designed to be used by widgets directly -- for example, plotly has an output renderer -- not as a layer on top of another widget framework. It also can't be used to just render HTML output directly, which is what a lot of R code does.

The approach taken here largely supersedes the notebook output approach (for console sessions). It extends our existing help proxy with a new file proxy. R HTML content is rendered to a temporary path on disk, and then we create a proxy that serves content from the path on disk. 

```mermaid
graph TD
proxy[Positron Proxy] --HTML --> pos[Positron]
html[(HTML)] -- read by--> proxy
pos -- create proxy --> proxy
R(R runtime) -- generates --> html
R -- show_html_file UI comm --> pos
```

Step by step:

1. An HTML widget is emitted in R.
2. Ark's custom HTML widget printer is invoked.
3. Ark renders the widget to a temporary directory, using `html_print`. 
4. Ark emits a `show_html_file` event over the UI comm to Positron with the path to the rendered widget.
5. Positron receives the `show_html_file` event.
6. Positron asks the Positron Proxy extension to create a server URL for the file path.
7. The Positron Proxy extension uses an Express server to create a route that serves the static content at a random URL. 
8. The Positron Proxy extension sends the URL back to Positron.
9. Positron opens the URL in the Viewer or Plots pane, as appropriate.

### New Features 

In addition to fixing some bugs around HTML rendering, we have a few new features enabled by this approach:

#### Popout Interactive Content

Because we now have a real URL for HTML content, it can be opened in web browsers. Plots that are based on HTML content now have a popout affordance:

![image](https://github.com/user-attachments/assets/67fda7d4-ddb0-46f3-b182-37bb698f76cf)

And so does HTML content rendered into the Viewer pane:

![image](https://github.com/user-attachments/assets/5271401d-822d-4f0a-b7ab-e9ebc94821f6)

#### For the Plots Pane haters

Don't like HTML going to your Plots pane? You can turn it off, so all HTML content goes to the Viewer pane.

<img width="550" alt="image" src="https://github.com/user-attachments/assets/30a2c79f-66ba-4924-b6f7-83b7367108b0">

*Why would you want to live this way, though?*

#### RStudio Viewer API

The `rstudio::viewer()` API is implemented with this change, so R packages that depend on it to show content will work correctly.

#### Open in Viewer context menu

You can now open HTML files directly in the Viewer pane using a context menu action in the Explorer pane:

<img width="422" alt="image" src="https://github.com/user-attachments/assets/0e269f6a-02fb-4384-94a5-5396f23cde44">

### QA Notes

Here's a bunch of test cases. These are all things that didn't work before this PR.

#### Mapview

Opens up a nice map in the Viewer pane. 

```r
library(mapview)
m <- mapview()
m
```

### ggiraph

An interactive 2D plot. Should open up in the Plots pane.


```r
library(ggplot2)
library(ggiraph)
g <- ggplot(mpg, aes(x = displ, y = cty)) +
geom_point_interactive(
aes(tooltip = model, data_id = model),
size = 3, hover_nearest = TRUE
)
girafe(ggobj = g)
```

#### Threejs

An interactive 3D plot. Should open up in the Plots pane.

```r
library(threejs)
z <- seq(-10, 10, 0.01)
x <- cos(z)
y <- sin(z)
scatterplot3js(x,y,z, color=rainbow(length(z)))
```

#### Model Summary

This is one of those packages that calls `rstudioapi::viewer` directly. 

```r
library(palmerpenguins)
library(fixest)
library(modelsummary)
m1 = feols(body_mass_g ~ bill_depth_mm + bill_length_mm | species, data = penguins)
modelsummary(m1)
```

#### React Table

Also should open in the Viewer.

```
library(reactable)
mtcars |> reactable::reactable()
```

#### Reprex

A common R package used to create reproducible examples. Should open in the Viewer.

```r
reprex::reprex({
  x <- rnorm(100)
  plot(x, sin(x))  
})
```

### Impacted Areas

This change touches code outside R HTML widgets, so these code paths should also be spot checked for issues.

- Previewing URLs in the Viewer (e.g. Streamlit, Shiny content)
- Previewing extension-provided content in the Viewer (e.g. Quarto)
- Notebook outputs in the Plots pane (e.g. IPyWidgets)

jmcphers committed:
add event to show HTML files

jmcphers committed:
add height parameter to event

jmcphers committed:
require integer height

jmcphers committed:
plumbing for show HTML file event

jmcphers committed:
HTML proxy stubs

jmcphers committed:
add HTML folder serving to PositronProxy

jmcphers committed:
wrap and rebroadcast html file events

jmcphers committed:
factor the previewer so 'html preview' is a sibling of 'url preview'

jmcphers committed:
add html file preview handler to service

jmcphers committed:
add webview passthroughs; make resolution optional

jmcphers committed:
create a Preview HTML rather than Preview URL

jmcphers committed:
custom action bar for HTML preview

jmcphers committed:
begin handling messages in plots service; add load indicator

jmcphers committed:
load plots into plots pane

jmcphers committed:
add an action to pop out the current plot

jmcphers committed:
add Positron API for opening an HTML file

jmcphers committed:
implement Open in Viewer command for HTML files

jmcphers committed:
raise plots pane for webview plots

jmcphers committed:
clean up proxy-backed plots when extension host stops

jmcphers committed:
handle extension host shutdown in Viewer pane

jmcphers committed:
show title of content in Viewer pane

jmcphers committed:
make more room for document title

jmcphers committed:
don't set a blank title

jmcphers committed:
allow folks to dump everything into Viewer if they want to

jmcphers committed:
cleanup/comments (round 1)

jmcphers committed:
Merge remote-tracking branch 'origin/main' into feature/html-widget-files

jmcphers committed:
DRY in plots service; other cleanup

jmcphers committed:
make interface to preview service more rational

jmcphers committed:
'title' is a better name for the freeform text field

softwarenerd commented on src/vs/workbench/contrib/positronPlots/browser/positronPlotsService.ts:
```diff
@@ -164,6 +183,34 @@ export class PositronPlotsService extends Disposable implements IPositronPlotsSe
+				return;
+			}
+			let removedSelectedPlot = false;
+			this._plots.forEach((plot) => {
```
This can be:

```JavaScript
this._plots.forEach((plot, index) => {
```

Which will remove the need to call:

```JavaScript
    this._plots.indexOf(plot)
```

On line 202.

softwarenerd commented on src/vs/workbench/contrib/positronPlots/browser/positronPlotsService.ts:
```diff
@@ -164,6 +183,34 @@ export class PositronPlotsService extends Disposable implements IPositronPlotsSe
+					if (this._selectedPlotId === plot.id) {
+						removedSelectedPlot = true;
+					}
+					this._plots.splice(this._plots.indexOf(plot), 1);
```
Index can be obtained by refactoring `forEach`, above.

softwarenerd commented on src/vs/workbench/contrib/positronPlots/browser/positronPlotsService.ts:
```diff
@@ -164,6 +183,34 @@ export class PositronPlotsService extends Disposable implements IPositronPlotsSe
+			// If we removed the selected plot, select the first plot in the list
+			if (removedSelectedPlot && this._plots.length > 0) {
+				this.selectPlot(this._plots[0].id);
+			}
```
I wonder whether we should select the nearest plot to the selected plot that was removed?

jmcphers replied:
A good question ... I think that'd be better but would make this a lot more complicated. Practically speaking I don't think this will happen much as the extension host doesn't get stopped (other than for a full shutdown) during the normal course of operation; this is really to handle extension host crashes, etc.

softwarenerd commented on src/vs/workbench/contrib/positronPreview/browser/components/actionBars.ts:
```diff
@@ -0,0 +1,34 @@
+ * PreviewActionBarsProps interface.
+ */
+export interface PreviewActionBarsProps extends PositronSessionsServices {
+	readonly commandService: ICommandService;
```
`commandService` is already in `PositronActionBarServices`. It is not needed in this interface declaration.

softwarenerd commented on src/vs/workbench/contrib/positronPreview/browser/components/actionBars.ts:
```diff
@@ -0,0 +1,34 @@
+ */
+export interface PreviewActionBarsProps extends PositronSessionsServices {
+	readonly commandService: ICommandService;
+	readonly configurationService: IConfigurationService;
```
`configurationService` is already in `PositronActionBarServices`. It is not needed in this interface declaration.

softwarenerd commented on src/vs/workbench/contrib/positronPreview/browser/components/actionBars.ts:
```diff
@@ -0,0 +1,34 @@
+export interface PreviewActionBarsProps extends PositronSessionsServices {
+	readonly commandService: ICommandService;
+	readonly configurationService: IConfigurationService;
+	readonly contextKeyService: IContextKeyService;
```
`contextKeyService` is already in `PositronActionBarServices`. It is not needed in this interface declaration.

softwarenerd commented on src/vs/workbench/contrib/positronPreview/browser/components/actionBars.ts:
```diff
@@ -0,0 +1,34 @@
+	readonly commandService: ICommandService;
+	readonly configurationService: IConfigurationService;
+	readonly contextKeyService: IContextKeyService;
+	readonly contextMenuService: IContextMenuService;
```
`contextMenuService` is already in `PositronActionBarServices`. It is not needed in this interface declaration.

softwarenerd commented on src/vs/workbench/contrib/positronPreview/browser/components/actionBars.ts:
```diff
@@ -0,0 +1,34 @@
+	readonly configurationService: IConfigurationService;
+	readonly contextKeyService: IContextKeyService;
+	readonly contextMenuService: IContextMenuService;
+	readonly keybindingService: IKeybindingService;
```
`keybindingService` is already in `PositronActionBarServices`. It is not needed in this interface declaration.

softwarenerd commented on src/vs/workbench/contrib/positronPreview/browser/components/htmlActionBars.tsx:
```diff
@@ -0,0 +1,103 @@
+
+export const HtmlActionBars = (props: PropsWithChildren<HtmlActionBarsProps>) => {
+
+	const [title, setTitle] = React.useState<string>(props.preview.html.title);
```
Nits:

I'd add `useState` to:

```
import { PropsWithChildren, useEffect } from 'react'; // eslint-disable-line no-duplicate-imports
```

Since `useEffect` is already there.

I'd remove the `<string>` as this type can be inferred.

softwarenerd commented on src/vs/workbench/contrib/positronPreview/browser/components/htmlActionBars.tsx:
```diff
@@ -0,0 +1,103 @@
+	};
+
+	// Main use effect.
+	useEffect(() => {
```
This `useEffect` will run on every render, continuously adding and removing the `onDidLoad` event handler.

softwarenerd commented on src/vs/workbench/contrib/positronPreview/browser/positronPreviewServiceImpl.ts:
```diff
@@ -174,34 +222,112 @@ export class PositronPreviewService extends Disposable implements IPositronPrevi
-				this._items.delete(key);
-			}
+		this._items.forEach((value, _key) => {
+			value.dispose();
```
Can be simplified to:

```
		this._items.forEach(value => {
			value.dispose();
		});
```

softwarenerd commented on src/vs/workbench/services/languageRuntime/common/languageRuntimeUiClient.ts:
```diff
@@ -73,6 +85,13 @@ export class UiClientInstance extends Disposable {
+	onDidShowHtmlFile: Event<IShowHtmlUriEvent>;
+
+	/** Emitter wrapper for Show URL events */
+	private _onDidShowUrlEmitter = this._register(new Emitter<ShowUrlEvent>());
```
I'd add `readonly`.

softwarenerd commented on src/vs/workbench/services/languageRuntime/common/languageRuntimeUiClient.ts:
```diff
@@ -73,6 +85,13 @@ export class UiClientInstance extends Disposable {
+	private _onDidShowUrlEmitter = this._register(new Emitter<ShowUrlEvent>());
+
+	/** Emitter wrapper for Show HTML File events */
+	private _onDidShowHtmlFileEmitter = this._register(new Emitter<IShowHtmlUriEvent>());
```
I'd add `readonly`.

softwarenerd reviewed (dismissed):
Looks very good. I made some comments.

jmcphers committed:
code review feedback

jmcphers commented:
@softwarenerd very much appreciate the review! Feedback should all be addressed in https://github.com/posit-dev/positron/pull/4151/commits/c5ce8648b1b7cf8b0b3b9db90999f20d533815a9.

jmcphers committed:
add editor title action for HTML files, too

DavisVaughan commented:
I tried the QA examples and they are working for me! (yay reprex!)

The only one not working is profvis, mentioned here: https://github.com/posit-dev/ark/pull/448#issuecomment-2253254947

malvezsilva commented:
The mentioned example of mapview is not working with version [2024.07.0-100](https://github.com/posit-dev/positron/releases/tag/2024.07.0-100), which is the most recent version I can see now.

```
library(mapview)
m <- mapview()
m

```

juliasilge commented:
@malvezsilva this is a PR and is not merged yet. Stay tuned!

malvezsilva commented:
Thanks!

El vie, 26 jul 2024 a las 15:58, Julia Silge ***@***.***>)
escribió:

> @malvezsilva <https://github.com/malvezsilva> this is a PR and is not
> merged yet. Stay tuned!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/posit-dev/positron/pull/4151#issuecomment-2253624909>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASEYSJG2HDGO7UL27LL2FBDZOLIA3AVCNFSM6AAAAABLOZID7WVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDENJTGYZDIOJQHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

jmcphers committed:
bump ark to 0.1.119

peekxc commented:
Does this PR also address #1924 (or #4118)?

jmcphers commented:
@peekxc Unfortunately not; currently this just addresses some issues with R HTML widgets. We're actively working on making HTML output better for Python, too, though; note that #4118 is currently planned to be addressed soon!
